pil-analyzer/src/expressionizer.rs

```diff
@@ -0,0 +1,403 @@
+                    "Converting builtin functions to expressions not supported.".to_string(),
+                ))
+            }
+            Value::Expression(e) => self.try_algebraic_expression_to_expression(e)?,
+        })
+    }
+
+    fn try_algebraic_expression_to_expression<T: FieldElement>(
+        &self,
+        e: &AlgebraicExpression<T>,
+    ) -> Result<Expression, EvalError> {
+        Ok(match e {
+            AlgebraicExpression::Reference(reference) => {
+                self.algebraic_reference_to_expression(reference)
+            }
+
+            AlgebraicExpression::BinaryOperation(AlgebraicBinaryOperation { left, op, right }) => {
+                BinaryOperation {
+                    left: Box::new(self.try_algebraic_expression_to_expression(left)?),
+                    op: (*op).into(),
+                    right: Box::new(self.try_algebraic_expression_to_expression(right)?),
+                }
+                .into()
+            }
+
+            AlgebraicExpression::UnaryOperation(AlgebraicUnaryOperation { op, expr }) => {
+                UnaryOperation {
+                    op: (*op).into(),
+                    expr: Box::new(self.try_algebraic_expression_to_expression(expr)?),
+                }
+                .into()
+            }
+
+            AlgebraicExpression::Challenge(Challenge { id, stage }) => {
+                let function = Expression::Reference(
+                    SourceRef::unknown(),
+                    Reference::Poly(PolynomialReference {
+                        name: "std::prelude::challenge".to_string(),
+                        type_args: None,
+                    }),
+                )
+                .into();
+                let arguments = [*stage as u64, *id]
+                    .into_iter()
+                    .map(|x| BigUint::from(x).into())
+                    .collect();
+                Expression::FunctionCall(
+                    SourceRef::unknown(),
+                    FunctionCall {
+                        function,
+                        arguments,
+                    },
+                )
+            }
+
+            AlgebraicExpression::Number(n) => Number {
+                value: n.to_arbitrary_integer(),
+                type_: Some(Type::Expr),
+            }
+            .into(),
+
+            AlgebraicExpression::PublicReference(s) => {
+                Expression::PublicReference(SourceRef::unknown(), s.clone())
+            }
+        })
+    }
+
+    fn algebraic_reference_to_expression(
+        &self,
+        AlgebraicReference {
+            name,
+            poly_id,
+            next,
+        }: &AlgebraicReference,
+    ) -> Expression {
+        // First we need to find out if the reference is an array element
+        // or a single symbol.
+
+        let ((_, array_start), symbol_name) = self
+            .poly_id_to_name
+            .range(..=(poly_id.ptype, poly_id.id))
+            .last()
+            .unwrap();
```

Comment:
Do I understand this right:
- `poly_id_to_name` would have an entry for the array, but not each element
- This code finds it's `PolyID.id` (which would that of the first element), assuming a consecutive ordering of the IDs of an array.

I think this deserves a comment :)